### PR TITLE
Added Virtual-Device for use with VSC-Design-Editor

### PR DESCRIPTION
**Pull request by DohyungLim**
- Connect to websocket server in Preview mode.
- Ignore if websocket server is not running.
- Added websocket connection part to VSC-Design-Editor Git.

Signed-off-by: DoHyung Lim <delight.lim@samsung.com>

### CODE REVIEW

**Comment by DohyungLim** [comment]
Currently only the Dimmer-Graph-Input (text, slider) widget is applied.
I will update it by adding a list later.

File: design-editor/src/panel/preview/preview-element-vd.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
+/* eslint-disable no-console */
+/* eslint-disable max-len */
+function startVD (profile) {
+	let stDevice, stProperty, stValue, eventType;
+	const previewEl = document.querySelector('.closet-preview-frame').contentWindow;
+	const tau = previewEl.tau;
+
+	/* iFrame Element */
+	const textData = previewEl.document.querySelector('input[type=\'text\']');
```
Comment (grzegorz-czajkowski) [inline design-editor/src/panel/preview/preview-element-vd.js]: Why we take care of first input element only? What if there are multiple elements?

File: design-editor/src/panel/preview/preview-element-vd.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		console.log('WebSocket is connected.');
+	};
+
+	connection.onerror = function (error) {
+		console.log(`not connect or server is down, ${error}`);
+	};
+
+	connection.onmessage = function (message) {
+		const json = JSON.parse(message.data);
+		const ElementArray = [];
```
Comment (grzegorz-czajkowski) [inline design-editor/src/panel/preview/preview-element-vd.js]: elementsArray ?

File: design-editor/src/panel/preview/preview-element-vd.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		if (json.type === 'oic.r.light.dimming') {
+			stDevice = json.type;
+			stProperty = json.data.text;
+			stValue = json.data.id;
+			console.log(`stDevice : ${stDevice} / stProperty : ${stProperty} / stValue : ${stValue}`);
+
+			if (textData != undefined || textData != null) {
```
Comment (grzegorz-czajkowski) [inline design-editor/src/panel/preview/preview-element-vd.js]: if (textData) {} is not sufficient?

File: design-editor/src/panel/preview/preview-element.js [hunk 4]
```diff
@@ -65,17 +66,18 @@ class Preview extends DressElement {
 		this.createFromTemplate(TEMPLATE_PATH, {
 			callback: () => {
 				const $frame = this.$el.find('.closet-preview-frame');
+				const screenConfig = StateManager.get('screen', {});
```
Comment (grzegorz-czajkowski) [inline design-editor/src/panel/preview/preview-element.js]: Wouldn't be better to pass this directly to startVD?

File: design-editor/src/panel/preview/preview-element-vd.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+	if (sliderData != undefined || sliderData != null) {
+		sliderData.addEventListener(eventType, (event) => {
+			const data = event.target.value;
+			const dataInfo = {
+				device: 'oic.r.light.dimming',
+				property: 'dimmingSetting',
+				value: data
+			};
+			if (textData != undefined || textData != null) {
```
Comment (grzegorz-czajkowski) [inline design-editor/src/panel/preview/preview-element-vd.js]: Why on change slider, we always update textData? Are they connected somehow?

**Review by grzegorz-czajkowski — COMMENTED** [review]
Hello @DohyungLim 
I checked this PR together with its dependency to VSC. It works ok for basic widgets. 

Please have in mind that this is already implemented by Sthings.js and iotivity.js without cloud as well (as mockup implementation). According to my check onSubscribe and send functions are working fine even if your PR is applied which means this is code and functionality duplication.

I'd prefer extending SThings and iotivity functionality to work with device launcher. What do you think?

File: design-editor/src/panel/preview/preview-element-vd.js (hunk 1 above)
Comment (DohyungLim) [reply design-editor/src/panel/preview/preview-element-vd.js]: We implemented the implementation based on TC.
TC adds Dimmer, Graph, and Input (text or slider or both) one by one to test the operation.
We plan to modify it to control several elements later.

File: design-editor/src/panel/preview/preview-element-vd.js (hunk 2 above)
Comment (DohyungLim) [reply design-editor/src/panel/preview/preview-element-vd.js]: yes, I will modify array name.

done.

File: design-editor/src/panel/preview/preview-element-vd.js (hunk 5 above)
Comment (DohyungLim) [reply design-editor/src/panel/preview/preview-element-vd.js]: Yes. There is a Slider or Text to confirm the change in the dimmer.
Confirm the change if a Slider or Text or both are added.

**Review by lmslachciak — COMMENTED** [review]
We have doubts regarding partial duplication of SThings functionality in this PR.
Please consider removing Sthings integration with DE in separate PR.

**Comment by DohyungLim** [comment]
@lslachciak 
We are preparing to build our app using the SmartThings CLI for the M2 release.
However, the CLI cannot be used.
So I want to tell the tester that Preview and mockup are working only with the parameters passed through DE for mockup.

The st-device and st-property may change the framework to disable it.
The current patch will provide testers with the integration of the Dimmer-Graph-Input widget.

In the future, we will make modifications to work with sthings.js through new PR.

**Comment by lmslachciak** [comment]
Converted PR to draft as it is suspended now

**Comment by lmslachciak** [comment]
@DohyungLim how about closing this PR if no intention to merge it?
I wish to do some cleanup in repos...

**Comment by lmslachciak** [comment]
I'm closing this PR to make some cleanup. Please reopen if needed.

**Pull request closed without merging**